<a href="https://colab.research.google.com/github/gmihran/AyrikSistemlerIleriOlasilik/blob/master/GozdeAltinsoyOdev2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Karşılaşma Problemi (Permütasyon)

### **Problem** : Bir spor salonuna günde 3000 kişi spor yapmak için geliyor. Gelen kişilerin %40'ı kilo vermek istiyor. Sabah saatlerinde (08:00-12:00 saatleri arasında) kulübe 200 kişi geliyor. 
### Sabah saatlerinde kulübe gelen kişilerden kilo vermek isteyen bir kişiyle karşılaşma olasılığımız nedir?

**p (Toplam Kişi)** = 3000

**s (Sabah saatlerinde gelen kişi sayısı)** = 200

**f (Kilo Vermek İsteyen Kişi Sayısı)** = 3000 * 40 / 100 = 1200

*Her kişi ile karşılaşma eşit şanslı olay olursa:*

**P (kilo vermek isten kişi sayısı)** = 1 - C (p-f,s) / C (p,s)

In [49]:
print("Kilo vermek isteyen kişi sayısı:",3000 * 40 / 100)

Kilo vermek isteyen kişi sayısı: 1200.0


In [0]:
def faktoriyel(sayi):
    if sayi==1:
        return 1
    else:
        return sayi * faktoriyel(sayi-1) 


def kombinasyon(n,r):
  return faktoriyel(n)/(faktoriyel(n-r)*faktoriyel(r))

In [51]:
p=3000
s=50
f=1200
k=p-f

P=1-kombinasyon(k,s)/kombinasyon(p,s)
print(P)

RecursionError: ignored

Recursive fonksiyon ile ilgili `RecursionError: maximum recursion depth exceeded in comparison` yani `Özyineleme Hatası: maksimum özyineleme derinliği karşılaştırmaya göre aşıldı` hatası almaktayız. Bu yüzden recursive fonksiyon yerine farklı bir yöntemle faktöriyel hesaplaması yapmalıyız.

Bunun için aşağıdaki faktöriyel hesaplamasını yapan fonksiyonu işlemi tekrar denemekteyiz.

In [0]:
def faktoriyel(sayi):
  sonuc=1
  for i in range(1,sayi+1):
    sonuc=sonuc*i
  return sonuc

In [53]:
p=3000
s=50
f=1200
k=p-f

P=1-kombinasyon(k,s)/kombinasyon(p,s)
print(P)

0.9999999999938683


### Sonuca bu fonksiyonlar ile ulaştık. Ama bu işlemlerdeki değerlere sadeleştirme işlemi uygulayarak sonuca daha hızlı bir şekilde ulaşabiliriz.
Sadeleştirme işlemi yapılarak daha kısa sürede faktöriyel hesaplanabilir.

**Bu adımlar aşağıdaki gibidir;**

K(n,r) kombinasyon işlemindeki n ile n-r değerleri karşılaştırırız. Bu değerlerden büyük olan değeri bölen değerle sadeleştirdikten sonra çarpım sonucunu hesaplatacağım değerleri faktöriyel fonksiyonuna gönderiririz. Sonraki adımda ise bir önceki adımda bulmuş olduğumuz değer ile küçük olan sayının faktöriyelini birbirine böleriz.

**Ör: K(8,2) değerini ele alalım.** 

**Adım-1:** 8 ile 6 sayısını karşılaştırdığımızda 6 değeri daha büyük olduğu için sadeleştirme işleminde bu iki sayıyı kullanırız.

**Adım-2:** 8! ile 6! sayılarını sadeleştirdiğimizde elimizde 7*8 değeri kalmaktadır.Faktöriyele bu değerlerin başlangıç ve son değerini gönderdiğimizde bize bu sayıların çarpımını geri döndürür.

**Adım-3:** 2! değerini hesaplatıp 2.adımda bulmuş olduğumuz 56 sonucunu 2 değerine böleriz. Böylelikle 28 değerine ulaşmış oluruz.

In [0]:
def fakt(n,r):
  sonuc=1
  for i in range(r+1,n+1):
  #r+1 den n değerine kadar faktöriyelini hesaplarız
    sonuc=sonuc*i
  return sonuc

def komb(n,r):
  nr=n-r
  if nr>r:
    K=fakt(n,nr)
    K=K/fakt(r,1)
  else:
    K=fakt(n,r)
    K=K/fakt(nr,1)
  return K

***Sağlama yapacak olursak;***

In [56]:
print(komb(8,2))

28.0


In [57]:
p=3000
s=50
f=1200
k=p-f

P=1-komb(k,s)/komb(p,s)
print(P)


0.9999999999938683


**Sonuç:** Sabah saatlerinde kulübe gelen kişilerden kilo vermek isteyen bir kişiyle karşılaşma olasılığımız **0.9999999999938683** olarak bulunur.